### PR TITLE
Validate registrations

### PR DESCRIPTION
**Pull request by gsvr**

### CODE REVIEW

**Comment by gsvr** [comment]
@imsickofmaps Please review. Let me know whether I should remove the postbirth validation or just leave it in.

File: familyconnect_registration/settings.py [hunk 1]
```diff
@@ -147,6 +148,8 @@
     'registrations.tasks',
 )
 
+CELERY_CREATE_MISSING_QUEUES = True
+
```
Comment (gsvr) [inline familyconnect_registration/settings.py:160]: Should I add a similar queue scheme as for ndoh?

Comment (imsickofmaps) [inline familyconnect_registration/settings.py:160]: I think we've learnt that being explicit saves stress later. So yes.

File: registrations/tasks.py [hunk 2]
```diff
@@ -1,23 +1,225 @@
...
+        if (registration.stage == "prebirth" and
+                registration.source.authority in ["hw_limited", "hw_full"] and
+                set(hw_pre_id).issubset(data_fields)):  # ignore extra data
+            if self.check_field_values(hw_pre_id, registration.data):
+                registration.data["reg_type"] = "hw_pre_id"
+                registration.data["preg_week"] = calc_pregnancy_week_lmp(
+                    get_today(), registration.data["last_period_date"])
+                registration.validated = True
+                registration.save()
+                return "Success"
```
Comment (imsickofmaps) [inline registrations/tasks.py]: What's the reason you're returning a string not boolean?

File: registrations/tasks.py [hunk 3]
```diff
@@ -1,23 +1,225 @@
...
+    return msg_receiver in ["head_of_household", "mother_to_be",
+                            "family_member", "trusted_friend"]
+
+
+def is_valid_loss_reason(loss_reason):
+    return loss_reason in ['miscarriage', 'stillborn', 'baby_died']
 
+
+def is_valid_name(name):
+    return isinstance(name, six.string_types)  # TODO reject non-letters
```
Comment (imsickofmaps) [inline registrations/tasks.py:47]: Can you open a new issues for TODO's? Maybe one for validations, one for calculations?

File: registrations/tasks.py [hunk 4]
```diff
@@ -1,23 +1,225 @@
...
     """
-    name = "familyconnect_registration.registrations.tasks.the_incr"
+    name = "familyconnect_registration.registrations.tasks.\
+    validate_registration"
 
-    def run(self, anum, **kwargs):
+    def check_field_values(self, fields, registration_data):
+        for field in fields:
+            if field in ["contact", "registered_by"]:
+                if not is_valid_uuid(registration_data[field]):
```
Comment (imsickofmaps) [inline registrations/tasks.py:81]: Not sure there's much point in doing this... A useful validation would be to actually validate it against the identity store.

Comment (imsickofmaps) [inline registrations/tasks.py:81]: Second thoughts, leave it for now

File: registrations/tasks.py [hunk 5]
```diff
@@ -1,23 +1,225 @@
...
+    """ Task to validate a registration model entry's registration
+    data.
     """
-    name = "familyconnect_registration.registrations.tasks.the_incr"
+    name = "familyconnect_registration.registrations.tasks.\
+    validate_registration"
 
-    def run(self, anum, **kwargs):
+    def check_field_values(self, fields, registration_data):
+        for field in fields:
```
Comment (imsickofmaps) [inline registrations/tasks.py:79]: Please set `failures = []` at the top of the checks and return that. Then push the validation failures into that list as they occur so all fields are checked and we end up with all failures logged in the model instance.

**Comment by imsickofmaps** [comment]
Generally :ok_hand: but refactor that validation and I'll re-review before +1

File: registrations/tasks.py (hunk 2 above)
Comment (gsvr) [inline registrations/tasks.py]: I use it in the `run` return string.

**Comment by gsvr** [comment]
@imsickofmaps Please recheck

**Comment by imsickofmaps** [comment]
:+1:

**Comment by imsickofmaps** [comment]
:+1:

**Pull request merged**